### PR TITLE
SPIREX-86 Implement a cache strategy for storing signature

### PR DESCRIPTION
**Pull request by willallves**

### CODE REVIEW

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
+package sigstorecache
+
+import (
+	"container/list"
+
+	"github.com/sigstore/cosign/pkg/oci"
+)
+
+const (
+	maximumAmountCache = 3
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go:10]: I would use a greater default value. Maybe 10 or 20?

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+type Cache interface {
+	GetSignature(key string) *Item
+	PutSignature(Item)
+}
+
+// cache implements Cache interface
+type Cacheimpl struct {
+	size  int
+	items *list.List
+	//mutex sync.RWMutex
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go:29]:
```suggestion
	// mutex sync.RWMutex
```

Comment (tjamir) [reply pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go:29]: Or remove this line.

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+// NewCache creates and returns a new cache
+func NewCache() Cache {
+	return &Cacheimpl{
+		size:  maximumAmountCache,
+		items: list.New(),
+		//mutex: sync.RWMutex{},
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go:37]: Same here, add space (due to lint reasons) or remove this line

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		{
+			name: "key1",
+			fields: fields{
+				size:  1,
+				items: cacheMemory.items,
+			},
+			args: args{
+				key: "key1",
+			},
+			want: cacheMemory.GetSignature(item1.Key),
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go:112]: I don't think this is the right way to do it. We are using the function under test in the "want" field. We should use a variable or constant here. I think we could use `item1.Value` here.

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Cacheimpl{
+				size:  tt.fields.size,
+				items: tt.fields.items,
+			}
+			c.PutSignature(tt.args.i)
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go:180]: We should add some kind of verification here, to check If the element put here was really put in the cache.

**Review by tjamir — CHANGES_REQUESTED** [review]

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+// getElement returns list element of an existing item
+func (c *Cacheimpl) getElement(key string) *list.Element {
+	for e := c.items.Front(); e != nil; e = e.Next() {
+		i := e.Value.(Item)
+		if i.Key == key {
+			return e
+		}
+	}
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go:84]: I think using a map here should be a better option for performance purposes

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+// Put puts a new item into the cache.
+// Put removes the least recently used item from the items list when the cache is full.
+// Put pushes the new item to the front of the items list to indicate that the new item is recently used.
+func (c *Cacheimpl) PutSignature(i Item) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	e := c.getElement(i.Key)
+	if e != nil {
+		c.items.Remove(c.items.Back())
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache.go:65]: Shouldn't be `c.items.Remove(e)` here?

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+									},
+									"type": "some type"
+								}
+								"optional": {
+									"subject": "spirex3@example.com"
+								}
+							}`),
+			},
+		},
+	}
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go:132]: To improve readability, we could declare those items in the begining of this file.

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+
+	item4 := cacheInstance.GetSignature(item3.Key)
+	if item4 != nil {
+		t.Error("a non-existing item's key should return a nil item")
+	}
+
+	item5 := cacheInstance.GetSignature(item1.Key)
+	if !reflect.DeepEqual(item5.Value, item1.Value) {
+		t.Error("an existing items key's should return the existing item")
+	}
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go:145]: We could use the same standard as other files here.

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+									},
+									"type": "some type"
+								}
+								"optional": {
+									"subject": "spirex4@example.com"
+								}
+							}`),
+			},
+		},
+	}
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go:233]: Same as above. We could reuse those variables declarations.

File: pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+	cacheInstance.PutSignature(item3)
+	if cacheInstance.items.Len() != 2 {
+		t.Error("item count should stay the same after putting a new key that overflows the cache", cacheInstance.items.Len())
+	}
+
+	cacheInstance.PutSignature(item4)
+	wantcached := cacheInstance.GetSignature(item4.Key)
+	if !reflect.DeepEqual(wantcached.Value, item4.Value) {
+		t.Error("an existing items key's should return the existing item")
+	}
```
Comment (tjamir) [inline pkg/agent/plugin/workloadattestor/k8s/sigstorecache/sigstorecache_test.go:255]: We could use the same standard as other files here.

**Review by tjamir — CHANGES_REQUESTED** [review]
There are some changes proposed here. As a suggestion, I submited a PR for this branch addressing the changes I propose here.

**Review by matheusdefarias — APPROVED** [review]

**Pull request merged**